examples/custom_rule.jl

```diff
@@ -0,0 +1,90 @@
+# differentiate, or we may have special higher-level knowledge about `f` that enables us to write a more 
+# efficient rule. So let's see how to write a custom rule for `f`.
+
+# ## Defining our first rule 
+
+# First, we import the functions [`EnzymeRules.forward`](@ref) and [`EnzymeRules.reverse`](@ref).
+# We will need to overload these functions in order to define our custom rules.
+
+import Enzyme.EnzymeRules: forward, reverse
+
+# In this section, we write a simple forward rule:
+
+function forward(func::Const{typeof(f)}, ::Type{<:Duplicated}, y::Duplicated, x::Duplicated)
+    println("Using custom rule!")
+    out = func.val(y.val, x.val)
+    y.dval .+= 2 .* x.val .* x.dval
+    return Duplicated(out, sum(y.dval)) 
+end
+
+# In the signature of our rule, we have made use of `Enzyme`'s activity annotations. Let's break down each one:
+# - the [`Const`](@ref) annotation on `f` indicates that we accept a function `f` that does not have a derivative component,
+#   which makes sense since `f` itself does not depend on any parameters.
+# - the [`Duplicated`](@ref) annotation given in the second argument annotates the return value of `f`. This means that
+#   our `forward` function should return an output of type `Duplicated`, containing the original output `sum(y)` and its derivative.
+# - the [`Duplicated`](@ref) annotations for `x` and `y` mean that our `forward` function handles inputs `x` and `y`
+#   which have been marked as `Duplicated`. We should update their shadows with their derivative contributions. 
+
+# In the logic of our forward function, we run the original function, populate `y.dval` (the shadow of `y`), 
+# and finally return a `Duplicated` for the output as promised. Let's see our rule in action! 
+# With the same setup as before:
+
+x  = [3.0, 1.0]
+dx = [1.0, 0.0]
+y  = [0.0, 0.0]
+dy = [0.0, 0.0]
+
+g(y, x) = f(y, x)^2 # function to differentiate
+
+@show autodiff(Forward, g, Duplicated(y, dy), Duplicated(x, dx)) # derivative of output w.r.t. x[1]
+@show dy; # derivative of y w.r.t. x[1]
+
+# We see that our custom forward rule has been triggered and gives the same answer as before.
+
+# ## A more comprehensive set of rules 
+
+# Our custom rule applies for the specific set of activities that are annotated for `f` in the above `autodiff` call. 
+# However, Enzyme has a number of other annotations. And of course, we'd like rules for reverse-mode too!. 
+# So let us define a more comprehensive set of custom rules for our function `f`.
+
+# If the output has a [`DuplicatedNoNeed`](@ref) annotation, it means we are only interested in its derivative.
+# To squeeze the last drop of performance, the below rule avoids computing the output of the original function and 
+# just computes its derivative.
+
+function forward(func::Const{typeof(f)}, ::Type{<:DuplicatedNoNeed}, y::Duplicated, x::Duplicated)
+    y.val .= x.val.^2 
+    y.dval .+= 2 .* x.val .* x.dval
+    return sum(y.dval)
+end
+
```

Comment:
if I keep going like this, I'm going to need a lot of rules 🙃 I don't know if there's an easy way of specifying these more compactly, say to at least get things correct for the case where `f` really needs a custom rule. (And then if the performance is desired maybe relying on some dead code elimination / thunking).

If that's not a thing yet, maybe for this tutorial I should just give a couple more rules for `f` to cover some representative cases and leave things there.

Reply:
So what you can do is not force the annotation on the type or return. And then can do if duplicated (compute return), else return just the derivative.

Reply:
(oops, was meant to include the TODO at the end in the code highlight here)

Reply:
is `Active` a third (non-batched) possibility for `RT`?

Reply:
Only if the return is a float or complex of float/etc.